sql/idempotent/005-vectorizer.sql

```diff
@@ -0,0 +1,1178 @@
+    if _item is null then
+        -- nothing to do
+        return;
+    end if;
+    -- execute the vectorizer
+    perform ai.execute_async_ext_vectorizer(_vectorizer_id);
+    commit;
+end
+$func$
+language plpgsql security invoker
+;
+
+-------------------------------------------------------------------------------
+-- _vectorizer_schedule_async_ext_job
+create or replace function ai._vectorizer_schedule_async_ext_job
+( vectorizer_id int
+, scheduling jsonb
+) returns bigint as
+$func$
+declare
+    _implementation text;
+    _sql text;
+    _extension_schema name;
+    _found bool;
+    _job_id bigint;
+begin
+    select pg_catalog.jsonb_extract_path_text(scheduling, 'implementation')
+    into strict _implementation
+    ;
+    case
+        when _implementation = 'timescaledb' then
+            -- look up schema/name of the extension for scheduling. may be null
+            select n.nspname into _extension_schema
+            from pg_catalog.pg_extension x
+            inner join pg_catalog.pg_namespace n on (x.extnamespace operator(pg_catalog.=) n.oid)
+            where x.extname operator(pg_catalog.=) _implementation
+            ;
+            if _extension_schema is null then
+                raise exception 'timescaledb extension not found';
+            end if;
+        when _implementation = 'pg_cron' then
+            select count(*) > 0 into strict _found
+            from pg_catalog.pg_extension x
+            where x.extname operator(pg_catalog.=) _implementation
+            ;
+            if not _found then
+                raise exception 'pg_cron extension not found';
+            end if;
+            _extension_schema = 'cron';
+        when _implementation = 'none' then
+            return null;
+        else
+            raise exception 'scheduling implementation not recognized';
+    end case;
+
+    -- schedule the job using the implementation chosen
+    case _implementation
+        when 'pg_cron' then
+            -- schedule the work proc with pg_cron
+            select pg_catalog.format
+            ( $$select %I.schedule(%L, %L, $sql$call ai._vectorizer_async_ext_job(null, pg_catalog.jsonb_build_object('vectorizer_id', %s))$sql$)$$
+            , _extension_schema
+            , pg_catalog.concat('vectorizer_', vectorizer_id)
+            , pg_catalog.jsonb_extract_path_text(scheduling, 'schedule')
+            , vectorizer_id
+            ) into strict _sql
+            ;
+            execute _sql into strict _job_id;
+        when 'timescaledb' then
+            -- schedule the work proc with timescaledb background jobs
+            select pg_catalog.format
+            ( $$select %I.add_job('ai._vectorizer_async_ext_job'::regproc, %s, config=>%L)$$
+            , _extension_schema
+            , ( -- gather up the arguments
+                select string_agg
+                ( case s.key
+                    when 'schedule_interval' then pg_catalog.format('%L', s.value)
```

Comment:
why can't we use `schedule_interval=>%L` here too. why do we. need the branch?

Reply:
just force of habit.

Reply:
0f087022f0ca17fbb466afc985aedc4a53b8faa3